### DISCUSSION TITLE
How to create an executable in OpenModelica ?

### DISCUSSION THREAD

**Discussion by ap766** [OpenModelica/OpenModelica#11294] [Q&A] [answered]
Basically the statement is that I should create an executable program of this model along with other dependent files that this executable program requires to execute it. It corresponds to the simulation of the mentioned model.

So I have created a simulation of the model however there is no .exe file created.When I click on export as you can see there is option of FMU File(pic 1)and I selected that.
And after I did that a folder , in this case - \NonInteractingTanks.TwoConnectedTanks got created at the C:\Users\lenovo\AppData\Local\Temp\OpenModelica\OMEdit\ directory.  And in pic 2 you can see the files .

I am supposed to execute the program from a pyqt6 app and I am not sure how to without the .exe.One of the things I need to do is allow the user to take input from the pyqt6 GUI for start and stop time and I see that start and stop time is in the XML file (of pic2)..but I am sure how to execute / create the .exe and then also take inputs from the pyqt6 application for start and stop time and create the executable according to that.    

Realy help in this.I am a beginner.

Pic 1
![image](https://github.com/OpenModelica/OpenModelica/assets/79255079/78663103-d361-43c2-8403-29f7e3de7b7a)

Pic 2
![image](https://github.com/OpenModelica/OpenModelica/assets/79255079/7add0d6b-ce3d-49f8-bff7-b67d58a938ac)

**Comment by ap766** [comment]
Using OMEdit for this.

**Comment by casella** [comment, marked as answer]
The exe file should be there. Can you enable showing file extensions in the File Explorer?

**Reply by ap766** [reply]
Yes, I wasnt able to find the .exe because there was no extensions enabled.

**Reply by casella** [reply]
Does it work?

**Reply by ap766** [reply]
Sorry for the late reply. Yes, when I go to that particular path.In this case -C:\Users\lenovo\AppData\Local\Temp\OpenModelica\OMEdit\NonInteractingTanks.TwoConnectedTanks and then execute TwoConnectedTanks.exe in the command line, it says simulation finished successfully. Although I am still trying to figure out to generate the plot too because OMPlot wasn't installed along with OpenModelica in my case.

**Comment by AnHeuermann** [comment]
If you want to build a tool on top of OpenModelica that will simulate a model I would recommend to use the Functional Mock-Up Interface (FMI) to export a Functional Mock-Up Unit (FMU). It is standardized and won't change (to drastically), while the "normal" model executable has no standardized interface and we could change it at some point.

The FMU won't contain an executable, but a shared library (.so or .dll) . How to interact with it is described in the the FMI standard: https://fmi-standard.org/
There are also other tools that can import FMUs. You might be able to find some inspiration in the open source tools listed in https://fmi-standard.org/tools/

-----

You can't run the model executable on Windows without setting some environment variables (`PATH` in particular). The batch script (*.bat) next to the executable will do this for a simulation call from OMEdit.

**Comment by ap766** [comment]
Thanks a lot for ypur help! I apologize for the late reply.Basically it goes like this:-

The task involves two parts. The first part is to compile a model and create an executable program from the OpenModelica simulation tool. The second part is to create a GUI application in Python that will run the above-generated executable with some parameters.

So for the first part, when I simulate using omedit , it creates a .exe file in  C:\Users\lenovo\AppData\Local\Temp\OpenModelica\OMEdit\NonInteractingTanks.TwoConnectedTank. And for the second path , I choose the .exe file in that directory nd execute it .And it says simulation finished successfully. Basically instead of directly going to that directory through the shell and typing TwoConnectedTanks.exe -override starttime=0,stoptime=5 , I do it through the GUI. (Altho plot isn't generated yet).

So you are suggesting to export as fmu ? how should I be executing the model in this case.

I think these are the basics, but I am actually new to this.So really appreciate your help.

**Reply by AnHeuermann** [reply]
We have some slides from last year's MODPROD Tutorial on FMI with OpenModelica and OMSimulator. Maybe those are a good starting point: https://github.com/OpenModelica/OMTeaching/tree/master/Tutorials/MODPROD22/Tutorial2

Also there is a section on FMI export in the [OpenModelica User's Guide](https://openmodelica.org/doc/OpenModelicaUsersGuide/latest/fmitlm.html#fmi-export).

On how to work with FMUs in detail I would suggest reading the [FMI standard](https://fmi-standard.org/). To get a feeling for it you can play around a bit with exporting FMUs and simulating them in different tools.

Implementing a standard compliant FMU importing tool from scratch is quiet some work. But you can probably build on top of already existing software (check the licenses on what you are allowed to do with the software you use).
Getting something to work with the executable only could be faster, but isn't documented and could break with every new release of OpenModelica.

**Comment by ap766** [comment]
Okay!thank you.
I am doing it with the executable for the time being.So just wanted to know if there is any specific command to generate the plot of the simulation?

**Reply by arun3688** [reply]
@ap766 If you have OMPlot installed then you can start the OMPlot from commandLine , see below

`>>> OMPlot --plotAll  `

the above command will show you list of result files and you can choose it from to plot the results

**Reply by ap766** [reply]
Thanks a lot! Is there any particular documentation I could refer to for all this?

**Reply by casella** [reply]
Have you tried the [User's Guide](https://openmodelica.org/doc/OpenModelicaUsersGuide/latest/index.html)?

**Reply by ap766** [reply]
Will look into it.Thank you
